src/nvim/drawline.c

```diff
@@ -2795,7 +2801,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, s
 
     // At end of screen line and there is more to come: Display the line
     // so far.  If there is no more to display it is caught above.
-    if (wlv.col >= grid->cols && (!has_fold || virt_line_offset >= 0)
+    if (wlv.col >= grid->cols && (!has_foldtext || virt_line_offset >= 0)
```

Comment:
You need to add conditions below so that a folded line behaves like a non-wrapped line.

Reply:
I've introduced:

```c
  const bool wrap = wp->w_p_wrap
                    && !has_fold;       // Never wrap folded lines
```

and replaced all used of `wp->w_p_wrap` with it.